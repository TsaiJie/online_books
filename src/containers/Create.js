import React, { PureComponent } from 'react'
import CategorySelect from '../components/CategorySelect'
import PriceForm from '../components/PriceForm'
import { Tab, Tabs } from '../components/Tabs'
import { testCategories } from '../testData'
import { TYPE_INCOME, TYPE_OUTCOME } from '../utility'
import { AppContext } from '../App'
import withContext from '../WithContext'
class Create extends PureComponent {
  render() {
    const { data } = this.props
    const filterCategories = testCategories.filter(
      (categroy) => categroy.type === TYPE_OUTCOME
    )
    return (
      <div
        className="create-page py-3 px-3 rounded mt-3"
        style={{ background: '#fff' }}
      >
        <Tabs activeIndex={0} onTabChange={() => {}}>
          <Tab>支出</Tab>
          <Tab>收入</Tab>
        </Tabs>
        <CategorySelect
          categories={filterCategories}
          onSelectCategory={() => {}}
        />
        <PriceForm
          onFormSubmit={(a, b) => {
            console.log(a, b)
          }}
          onCancelSubmit={(e) => {
            console.log(e)
          }}
        />
      </div>
    )
  }
}

export default withContext(Create)
