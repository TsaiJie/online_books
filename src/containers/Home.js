import React, { PureComponent } from 'react'
import PriceList from '../components/PriceList'
import Ionicon from 'react-ionicons'
import { AppContext } from '../App'
import {
  LIST_VIEW,
  CHART_VIEW,
  TYPE_INCOME,
  TYPE_OUTCOME,
  parseToYearAndMonth,
  padLeft,
} from '../utility'
import TotalPrice from '../components/TotalPrice'
import MonthPicker from '../components/MonthPicker'
import CreateBtn from '../components/CreateBtn'
import logo from '../logo.svg'
import { Tabs, Tab } from '../components/Tabs'
export const categories = {
  1: {
    id: '1',
    name: '旅行',
    type: 'outcome',
    iconName: 'ios-plane',
  },
  2: {
    id: '2',
    name: '理财',
    type: 'income',
    iconName: 'logo-yen',
  },
}
export const items = [
  {
    id: 1,
    title: '去云南旅游',
    date: '2020-12-10',
    price: 200,
    cid: 1,
  },
  {
    id: 2,
    title: '去云南旅游',
    date: '2020-12-1',
    price: 400,
    cid: 1,
  },
  {
    id: 3,
    title: '理财收入',
    date: '2020-12-22',
    price: 400,
    cid: 2,
  },
]
export const newItem = {
  id: 4,
  title: '新添加的项目',
  date: '2020-12-10',
  price: 300,
  cid: 1,
}
const tabsText = [LIST_VIEW, CHART_VIEW]
export default class Home extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      items,
      currentDate: parseToYearAndMonth('2020/12/10'),
      tabView: LIST_VIEW,
    }
  }
  changeView = (index) => {
    this.setState({ tabView: tabsText[index] })
  }
  changeDate = (year, month) => {
    this.setState({ currentDate: { year, month } })
  }
  modifyItem = (modifiedItem) => {
    const modifiedItems = this.state.items.map((item) => {
      if (item.id === modifiedItem.id) {
        return { ...item, title: '更新后的标题' }
      } else {
        return item
      }
    })
    this.setState({
      items: modifiedItems,
    })
  }
  createItem = () => {
    this.setState({
      items: [newItem, ...this.state.items],
    })
  }
  deleteItem = (deletedItem) => {
    const filteredItems = this.state.items.filter(
      (item) => item.id !== deletedItem.id
    )
    this.setState({
      items: filteredItems,
    })
  }

  render() {
    const { items, currentDate, tabView } = this.state
    const itemsWithCategory = items
      .map((item) => {
        item.category = categories[item.cid]
        return item
      })
      .filter((item) => {
        return item.date.includes(
          `${currentDate.year}-${padLeft(currentDate.month)}`
        )
      })
    let totalIncome = 0,
      totalOutcome = 0
    items.forEach((item) => {
      if (item.category.type === TYPE_OUTCOME) {
        totalOutcome += item.price
      } else {
        totalIncome += item.price
      }
    })
    return (
      <AppContext.Consumer>
        {({ state }) => {
          return (
            <React.Fragment>
              <header className="App-header">
                <div className="row mb-5 justify-content-center">
                  <img src={logo} className="App-logo" alt="logo" />
                </div>
                <div className="row">
                  <div className="col">
                    <MonthPicker
                      year={currentDate.year}
                      month={currentDate.month}
                      onChange={this.changeDate}
                    />
                  </div>
                  <div className="col">
                    <TotalPrice income={totalIncome} outcome={totalOutcome} />
                  </div>
                </div>
              </header>
              <div className="content-area py-3 px-3">
                <Tabs activeIndex={0} onTabChange={this.changeView}>
                  <Tab>
                    <Ionicon
                      className="rounded-circle mr-2"
                      fontSize="25px"
                      color={'#007bff'}
                      icon={'ios-paper'}
                    />
                    列表模式
                  </Tab>
                  <Tab>
                    <Ionicon
                      className="rounded-circle mr-2"
                      fontSize="25px"
                      color={'#007bff'}
                      icon={'ios-pie'}
                    />
                    图表模式
                  </Tab>
                </Tabs>
                <CreateBtn onClick={this.createItem} />
                {tabView === LIST_VIEW && (
                  <PriceList
                    items={itemsWithCategory}
                    onModifyItem={this.modifyItem}
                    onDeleteItem={this.deleteItem}
                  />
                )}
                {tabView === CHART_VIEW && (
                  <h1 className="chart-title">这里是图表区域</h1>
                )}
              </div>
            </React.Fragment>
          )
        }}
      </AppContext.Consumer>
    )
  }
}
