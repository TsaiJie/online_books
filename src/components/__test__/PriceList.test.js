import React from 'react'
import { shallow } from 'enzyme'
import PriceList from '../PriceList'
import Ionicon from 'react-ionicons'
import { items, categories } from '../../containers/Home'
const itemsWithCategory = items.map((item) => {
  item.category = categories[item.cid]
  return item
})
const props = {
  items: itemsWithCategory,
  onModifyItem: jest.fn(),
  onDeleteItem: jest.fn(),
}
let wrapper
describe('test PriceList component', () => {
  beforeEach(() => {
    wrapper = shallow(<PriceList {...props} />)

  })
  it('should render the component to match snapshot', () => {
    // wrapper.getElement() 修复empty ShallowWrapper {}
    expect(wrapper.getElement()).toMatchSnapshot()
  })
  it('should render correct price items length', () => {
    expect(wrapper.find('.list-group-item').length).toEqual(itemsWithCategory.length)
  })
  it('should render correct icon and price for each item', () => {
    // first()取第一个元素 props() 属性
    const iconList = wrapper.find('.list-group-item').first().find(Ionicon)
    // console.log(iconList.debug())
    expect(iconList.length).toEqual(3)
    expect(iconList.first().props().icon).toEqual(itemsWithCategory[0].category.iconName)
    // console.log(iconList.first().props())
  })
  it('should trigger the correct function callbacks', ()=>{
    const firstItem = wrapper.find('.list-group-item').first()
    firstItem.find('a').first().simulate('click')
    expect(props.onModifyItem).toHaveBeenCalledWith(itemsWithCategory[0])
    firstItem.find('a').last().simulate('click')
    expect(props.onDeleteItem).toHaveBeenCalledWith(itemsWithCategory[0])
  })

  
})
