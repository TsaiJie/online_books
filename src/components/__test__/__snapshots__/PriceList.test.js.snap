// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceList component should render the component to match snapshot 1`] = `
<ul
  className="list-group list-group-flush"
>
  <li
    className="list-group-item d-flex justify-content-between align-items-center"
  >
    <span
      className="col-1"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      去云南旅游
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      -
      200
      元
    </span>
    <span
      className="col-2"
    >
      2020-12-10
    </span>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "cursor": "pointer",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "cursor": "pointer",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
  <li
    className="list-group-item d-flex justify-content-between align-items-center"
  >
    <span
      className="col-1"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      去云南旅游
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      -
      400
      元
    </span>
    <span
      className="col-2"
    >
      2020-12-1
    </span>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "cursor": "pointer",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "cursor": "pointer",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
  <li
    className="list-group-item d-flex justify-content-between align-items-center"
  >
    <span
      className="col-1"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="logo-yen"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      理财收入
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      +
      400
      元
    </span>
    <span
      className="col-2"
    >
      2020-12-22
    </span>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "cursor": "pointer",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "cursor": "pointer",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
</ul>
`;
